scvi/external/cellassign/_module.py

```diff
@@ -0,0 +1,238 @@
+        # though we do need the base expression intercept either way
+        design_matrix_col_dim = n_batch + n_continuous_cov
+        design_matrix_col_dim += 0 if n_cats_per_cov is None else sum(n_cats_per_cov)
+        self.register_buffer("rho", rho)
+
+        # perform all other initialization
+        self.min_delta = 2
+        dirichlet_concentration = torch.tensor([1e-2] * self.n_labels)
+        self.register_buffer("dirichlet_concentration", dirichlet_concentration)
+        self.shrinkage = True
+        self.b_g_0 = torch.nn.Parameter(torch.randn(n_genes))
+
+        # compute theta
+        self.theta_logit = torch.nn.Parameter(torch.randn(self.n_labels))
+
+        # compute delta (cell type specific overexpression parameter)
+        self.delta_log_unclamped = torch.nn.Parameter(
+            torch.FloatTensor(self.n_genes, self.n_labels).uniform_(-2, 2)
+        )
+
+        # shrinkage prior on delta
+        if self.shrinkage:
+            self.delta_log_mean = torch.nn.Parameter(
+                torch.zeros(
+                    1,
+                )
+            )
+            self.delta_log_variance = torch.nn.Parameter(
+                torch.ones(
+                    1,
+                )
+            )
+
+        self.log_a = torch.nn.Parameter(torch.zeros(B, dtype=torch.float64))
+
+    def _get_inference_input(self, tensors):
+        return {}
+
+    def _get_generative_input(self, tensors, inference_outputs):
+        x = tensors[_CONSTANTS.X_KEY]
+        size_factor = tensors["_size_factor"]
+
+        to_cat = []
+        if self.n_batch > 0:
+            to_cat.append(one_hot(tensors[_CONSTANTS.BATCH_KEY], self.n_batch))
+
+        cont_key = _CONSTANTS.CONT_COVS_KEY
+        if cont_key in tensors.keys():
+            to_cat.append(tensors[cont_key])
+
+        cat_key = _CONSTANTS.CAT_COVS_KEY
+        if cat_key in tensors.keys():
+            for cat_input, n_cat in zip(
+                torch.split(tensors[cat_key], 1, dim=1), self.n_cats_per_cov
+            ):
+                to_cat.append(one_hot(cat_input, n_cat))
+
+        design_matrix = torch.cat(to_cat) if len(to_cat) > 0 else None
+
+        input_dict = dict(x=x, size_factor=size_factor, design_matrix=design_matrix)
+        return input_dict
+
+    @auto_move_data
+    def inference(self):
+        return {}
+
+    @auto_move_data
+    def generative(self, x, size_factor, design_matrix=None):
+        self.delta_log = torch.clamp(
+            self.delta_log_unclamped, min=np.log(self.min_delta)
+        )
+        # x has shape (n, g)
+        delta = torch.exp(self.delta_log)  # (g, c)
+        theta_log = F.log_softmax(self.theta_logit)  # (c)
+
+        # compute mean of NegBin - shape (n_cells, n_genes, n_labels)
+        s = size_factor
+        base_mean = torch.log(s)  # (n, 1)
+        base_mean_u = base_mean.unsqueeze(-1)  # (n, 1, 1)
+        base_mean_e = base_mean_u.expand(
+            s.shape[0], self.n_genes, self.n_labels
```

Comment:
Let's go for a round of cleanup:

`s.shape[0]` gets used a lot, you can instead define a variable near the top of the function, you'll have to check but I think doing like `n_cells = x.shape[0]` is the same